pkg/operator/controller/controller.go

```diff
@@ -399,9 +399,41 @@ func (r *reconciler) getClusterIPFromNetworkConfig() (string, error) {
 	if err != nil {
 		return "", fmt.Errorf("invalid service cidr %v: %v", serviceCIDR, err)
 	}
+	// Ensure dnsClusterIP is not allocated to an existing service.
+	if err := r.ensureServiceIP(dnsClusterIP); err != nil {
+		return "", fmt.Errorf("failed to ensure service IP %s: %v", dnsClusterIP, err)
+	}
 	return dnsClusterIP.String(), nil
 }
 
+// ensureServiceIP ensures addr is not assigned to an existing service,
+// returning an error if addr is already allocated or failures occur
+// while checking service resources for addr.
+func (r *reconciler) ensureServiceIP(addr net.IP) error {
+	namespaces := &corev1.NamespaceList{}
+	if err := r.client.List(context.TODO(), namespaces); err != nil {
+		return fmt.Errorf("failed to list namespaces: %v", err)
+	}
+	for _, ns := range namespaces.Items {
+		services := &corev1.ServiceList{}
+		if err := r.client.List(context.TODO(), services); err != nil {
+			return fmt.Errorf("failed to list services for namespace %s: %v", ns.Name, err)
+		}
+		for _, svc := range services.Items {
+			if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: svc.Namespace, Name: svc.Name}, &svc); err != nil {
```

Comment:
You already have the services; why get it again?

You need to ignore the DNS service itself so that you don't report that the service that _should_ have the address is in conflict.

Keep in mind that this logic runs every iteration, so it should avoid excessive API use.

Reply:
Commit `4ce15d5` addresses the feedback above.